infra/Pos/Slotting/Impl/Util.hs

```diff
@@ -17,55 +17,79 @@ import           Pos.Core.Types              (EpochIndex, SlotId (..), Timestamp
                                               mkLocalSlotIndex)
 import           Pos.Util.Util               (leftToPanic)
 
+import           Pos.Slotting.Util           (getEpochSlottingDataEmpatically)
 import           Pos.Slotting.MemState.Class (MonadSlotsData (..))
-import           Pos.Slotting.Types          (EpochSlottingData (..), SlottingData (..))
+import           Pos.Slotting.Types          (EpochSlottingData (..))
 
 -- | Approximate current slot using outdated slotting data.
-approxSlotUsingOutdated :: MonadSlotsData m => EpochIndex -> Timestamp -> m SlotId
+approxSlotUsingOutdated
+    :: (MonadSlotsData m)
+    => EpochIndex
+    -> Timestamp
+    -> m SlotId
 approxSlotUsingOutdated penult t = do
-    SlottingData {..} <- getSlottingData
-    systemStart <- getSystemStart
+    systemStart <- getSystemStartM
+    sdLast      <- getNextEpochSlottingDataM
     let epochStart = esdStartDiff sdLast `addTimeDiffToTimestamp` systemStart
     pure $
         if | t < epochStart -> SlotId (penult + 1) minBound
            | otherwise      -> outdatedEpoch systemStart t (penult + 1) sdLast
   where
     outdatedEpoch systemStart (Timestamp curTime) epoch EpochSlottingData {..} =
         let duration = convertUnit esdSlotDuration
-            start = getTimestamp (esdStartDiff `addTimeDiffToTimestamp` systemStart) in
+            start = getTimestamp (esdStartDiff `addTimeDiffToTimestamp` systemStart)
+        in
         unflattenSlotId $
         flattenEpochIndex epoch + fromIntegral ((curTime - start) `div` duration)
 
+-- TODO(ks): Move to MonadSlotsData?
```

Comment:
No, we want to get rid of `MonadSlotsData` because it's useless abstraction (it doesn't have multiple implementations, doesn't really hide anything, etc.). Sooner or later it will be replaced by `(MonadReader ctx m, HasSlottingVar ctx)`.